google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -278,6 +474,65 @@ def __init__(self, firewall_rules=None, firewall_rule_action='allow'):
 
         self._expanded_rules = None
 
+    def json_dict(self):
+        """Gets the JSON key and values for the firewall action.
+
+        Returns:
+          tuple: Of key ('allowed' or 'denied') and the firewall rules.
+
+        Raises:
+          InvalidFirewallActionError: If a rule is not formatted for the API.
+        """
+        self.validate()
+        if self.action == 'allow':
+            return ('allowed', self.rules)
+        return ('denied', self.rules)
+
+    # pylint: disable=too-many-branches
+    def validate(self):
+        """Validates that the firewall rules are valid for use in the API.
+
+        Raises:
+          InvalidFirewallActionError: If a rule is not formatted for the API.
+        """
+        for rule in self.rules:
+            if 'IPProtocol' not in rule:
+                raise InvalidFirewallActionError(
+                    'Action must have field IPProtocol')
+            if 'ports' in rule:
+                if rule['IPProtocol'] not in ['tcp', 'udp']:
+                    raise InvalidFirewallActionError(
+                        'Only "tcp" and "udp" can have ports specified: %s' %
+                        rule)
+                for port in rule['ports']:
+                    if '-' in port:
+                        split_ports = port.split('-')
+                        if len(split_ports) > 2:
+                            raise InvalidFirewallActionError(
+                                'Invalid port range: %s' % port)
+                        if int(split_ports[0]) < 0:
+                            raise InvalidFirewallActionError(
+                                'Port range must start > 0: %s' % port)
+                        if int(split_ports[1]) > 65535:
+                            raise InvalidFirewallActionError(
+                                'Port range must end < 65536: %s' % port)
+                        if int(split_ports[0]) > int(split_ports[1]):
+                            raise InvalidFirewallActionError(
+                                'Start port range > end port range: %s' % port)
+                    else:
+                        try:
+                            int(port)
```

Comment:
Also validate range of port, is >= 0 and <= 65535.

Reply:
Done.